File: utils/Concepts/sidef_executor.lua
#!/usr/bin/lua

-- Sidef executor in Lua

String = setmetatable({
    __index = function(self, i)
        return String[i]
    end,

    concat = function(self, arg)
        return String(self.value .. arg.value)
    end,

    say = function(self)
        self:concat(String("\n")):print();
    end,

    print = function(self)
        io.write(self.value);
    end,
},
{
    __call = function(self, val)
        return setmetatable({value = val}, String)
    end
});

function execute_expr(statement)

    local self_obj = statement['self'];

    if (type(self_obj) == "table" and getmetatable(self_obj) == nil) then
        self_obj = execute(self_obj);
    end

    if (statement['call'] ~= nil) then
        for i = 1, #(statement['call']) do

            local call = statement['call'][i];
            local meth = call['method'];

            if (call['arg'] ~= nil) then

                local args = {};
                local call_args = call['arg'];

                for i = 1, #call_args do
                    local arg = call_args[i];

                    if (type(arg) == "table" and getmetatable(arg) == nil) then
                        arg = execute_expr(arg);
                    end

                    args[i] = arg;
                end

                self_obj = self_obj[meth](self_obj, unpack(args));
            else
                self_obj = self_obj[meth](self_obj);
            end

        end
    end

    return self_obj;
end

function execute(structure)
    local results = {};

    for i = 1, #(structure['main']) do
        local statement = structure['main'][i];
        results[i] = execute_expr(statement);
    end

    return results[#results];
end

local ast = {
    main = {
        {
            call = {{method = "print"}},
            self = {
                main = {
                    {
                        call = {{method = "concat", arg = {{self = String("llo")}}}},
                        self = String("he"),
                    }
                },
            }
        },
        {
            call = {{method = "say"}},
            self = String(" world!");
        },
    }
}

execute(ast);

File: utils/Concepts/sidef_executor.jl
#!/usr/bin/julia

import Base.print;
import Base.+;

#
## Boolean type
#
mutable struct SBool
    value::Bool;

    function SBool(b::Bool)
        this = new();
        this.value = b;
        return this;
    end
end

#
## String type
#
mutable struct SString
    value::AbstractString;

    function SString(s::AbstractString)
        this = new();
        this.value = s;
        return this;
    end
end

#
## String methods
#
function say(s::SString)
    println(s.value);
    SBool(true);
end

function print(s::SString)
    print(s.value);
    SBool(true);
end

function +(a::SString, b::SString)
    SString(a.value * b.value);
end

#
## Expression executor
#
function execute_expr(statement)
    if haskey(statement, :self)
        self = statement[:self];

        if isa(self, Dict)
            self = execute(self);
        end

        if haskey(statement, :call)
            for call in statement[:call]
                method = Symbol(call[:method]);

                if haskey(call, :arg)

                    # Evaluate and collect the arguments
                    args = Any[];
                    for arg in call[:arg]
                        if isa(arg, Dict)
                            push!(args, execute_expr(arg))
                        else
                            push!(args, arg);
                        end
                    end

                    # Call method with arguments
                    self = eval(:($method($self, $args...)));
                else
                    # Call method without arguments
                    self = eval(:($method($self)));
                end
            end
        end

        return(self);
    end
end

#
## Parse-tree executor
#
function execute(structure)
    results = Any[];
    for statement in structure[:main]
        push!(results, execute_expr(statement));
    end
    results[end];
end

#
## The AST
#
structure = Dict{Symbol, Any}(
    :main => [
        Dict{Symbol, Any}(
            :call => [Dict{Symbol, Any}(:method => "print")],
            :self => Dict{Symbol, Any}(
                :main => [
                    Dict{Symbol, Any}(
                        :call => [Dict{Any, Any}(:method => "+", :arg => [Dict{Symbol, Any}(:self => SString("llo"))])],
                        :self => SString("he"),
                    )
                ],
            )
        ),
        Dict{Symbol, Any}(
            :call => [Dict{Symbol, Any}(:method => "say")],
            :self => SString(" world!");
        ),
    ]
);

#
## Begin execution
#
execute(structure);
